jax_distribution/systems/ippo_recurrent_anakin.py

```diff
@@ -0,0 +1,440 @@
+                    0,
+                ),
+                batch,
+            )
+
+            train_state, total_loss = jax.lax.scan(
+                _update_minbatch, train_state, minibatches
+            )
+            update_state = (
+                train_state,
+                init_hstate,
+                traj_batch,
+                advantages,
+                targets,
+                rng,
+            )
+            return update_state, total_loss
+
+        init_hstate = initial_hstate[:, None, :]  # TBH
+        update_state = (
+            train_state,
+            init_hstate,
+            traj_batch,
+            advantages,
+            targets,
+            rng,
+        )
+        update_state, loss_info = jax.lax.scan(
+            _update_epoch, update_state, None, config["UPDATE_EPOCHS"]
+        )
+        train_state = update_state[0]
+        metric = traj_batch.info
+        rng = update_state[-1]
+
+        runner_state = (train_state, env_state, last_obs, last_done, hstate, rng)
+        return runner_state, metric
+
+    def runner_fn(runner_state):
+        """Vectorise and repeat the update."""
+        batched_update_fn = jax.vmap(
+            _update_step, axis_name="batch"
+        )  # vectorize across batch.
+
+        runner_state, metric = jax.lax.scan(
+        batched_update_fn, runner_state, None, config["NUM_UPDATES"]
+        )
+        return {"runner_state": runner_state, "metric": metric}
+
+    return runner_fn
+
+def run_experiment(env_name, config):
+
+    # INIT ENV
+    # TODO: We need a more general way of doing this
+    # If else statements do not scale well
+    if "MultiCVRP" in env_name:
+        config["NUM_AGENTS"] = 3
+        generator = UniformRandomGenerator(
+            num_vehicles=config["NUM_AGENTS"], num_customers=6
+        )
+        env = jumanji.make(env_name, generator=generator)
+        num_actions = int(env.action_spec().maximum)
+
+    elif "RobotWarehouse" in env_name:
+        env = jumanji.make(env_name)
+        num_actions = int(env.action_spec().num_values[0])
+        config["NUM_AGENTS"] = env.num_agents
+
+    env = AutoResetWrapper(env)
+    env = LogWrapper(env)
+    cores_count = len(jax.devices())
+    # Number of updates
+    timesteps_per_iteration = (
+        cores_count * config["ROLLOUT_LENGTH"] * config["BATCH_SIZE"]
+    )
+    config["NUM_UPDATES"] = (
+        config["TOTAL_TIMESTEPS"] // timesteps_per_iteration
+    )
+
+   # INIT NETWORK AND OPTIMIZER
+    network = ActorCriticRNN(num_actions, config=config)
+    rng = jax.random.PRNGKey(config["SEED"])
+    rng, _rng = jax.random.split(rng)
+    init_x = (
+        process_observation(env.observation_spec().generate_value(), env_name),
+        jnp.zeros((1)),
+    )
+    init_hstate = ScannedRNN.initialize_carry(config["NUM_AGENTS"], 128)
+    
+    # Support for multiple agents
+    init_net_x, init_net_hstate = jax.tree_util.tree_map(
+        lambda x: x[None, ...],
```

Comment:
Why is this extra batch dimension needed? It is also used in the ff case. Is the observation not already batched by agents?